chain/ethereum/src/ethereum_adapter.rs

```diff
@@ -1895,3 +1892,125 @@ fn resolve_transaction_receipt(
         }
     }
 }
+
+/// Retrieves logs and the associated transact receipts, if required by the [`EthereumLogFilter`].
+async fn get_logs_and_transactions(
+    adapter: Arc<EthereumAdapter>,
+    logger: &Logger,
+    subgraph_metrics: Arc<SubgraphEthRpcMetrics>,
+    from: BlockNumber,
+    to: BlockNumber,
+    log_filter: EthereumLogFilter,
+    unified_api_version: &UnifiedMappingApiVersion,
+) -> Result<Vec<EthereumTrigger>, anyhow::Error> {
+    // obtain logs externally
+    let logs = adapter
+        .logs_in_block_range(logger, subgraph_metrics, from, to, log_filter.clone())
+        .await?;
+
+    // not all logs have associated transaction hashes, nor do all triggers require them.
+    // we also restrict receipts retrieval for some api versions.
+    let transaction_hashes_by_block: HashMap<H256, HashSet<H256>> = logs
+        .iter()
+        .filter(|_| unified_api_version.equal_or_greater_than(&API_VERSION_0_0_7))
+        .filter(|log| {
+            if let Some(signature) = log.topics.first() {
+                log_filter.requires_transaction_receipt(signature, Some(&log.address))
+            } else {
+                false
+            }
+        })
+        .filter_map(|log| {
+            if let (Some(block), Some(txn)) = (log.block_hash, log.transaction_hash) {
+                Some((block, txn))
+            } else {
+                // TODO: Should this case be considered an error?
+                None
```

Comment:
Did you figure this out? Why would it be?

Reply:
It seems the `None` variant might come from pending transactions, which we don't handle.

Here is an example where it's being unwrapped freely:
https://github.com/graphprotocol/graph-node/blob/852161f40c3670e0a08d9ea4e30084b927993d8c/chain/ethereum/src/trigger.rs#L268

I've updated the comment on a74dd686d to replace the `TODO` with a proper explanation.